### ISSUE TITLE
CA1508 should not warn when pattern matching with type parameter

### ISSUE THREAD

**Issue by JohanLarsson** [dotnet/roslyn-analyzers#4638]
```cs
static IEnumerable<T2> M<T1, T2>(this IEnumerable<T1> source)
{
    using var e = source.GetEnumerator();
    while (e.MoveNext())
    {
        if (e.Current is T2 x)
        {
            yield return x;
        }
    }
}
```

![image](https://user-images.githubusercontent.com/1640096/103353230-9f269880-4aa8-11eb-9a4f-f372c24e3439.png)

**Comment by bartdesmet** [comment]
Hit some instances of this today as well. Similar example:

```csharp
static void G<T>(Dictionary<T, bool> d)
{
    foreach (var kv in d)
    {
        var k = kv.Key;
        if (k is IFormattable f)
        {
            // CA1508 claims the check is always true
        }
    }
}
```

**Comment by mavasani** [comment]
Verified neither of the above cases repro on the latest .NET8 preview SDK
